### ISSUE TITLE
NullPointerException with two filterMany-statements in same query

### ISSUE THREAD

**Issue by pieterbos** [ebean-orm/avaje-ebeanorm#588]
I was working on a toy-project, to test the new features of ebean. And to check how the latest ebean compares to hibernate 5 as well.

I ended up with the following query:

```java
public MessageThread byIdFullyFetched(Long id, User user) {
        Query<User> nonBlockedUsers = User.find.where().select(User.find.where().id).blockedUsers.equalTo(user).query();
        Query<User> nonBlockedUsers2 = User.find.where().select(User.find.where().id).blockedUsers.equalTo(user).query();

        ExpressionList<Message> nonBlockedMessages = Message.find.where().getExpressionList().notIn("author.id", nonBlockedUsers);
        ExpressionList<Like> nonBlockedLikes = Like.find.where().getExpressionList().notIn("liker.id", nonBlockedUsers2);

        return where()
                .id.eq(id)
                .messages.filterMany(nonBlockedMessages)
                .messages.likes.filterMany(nonBlockedLikes)
                .findUnique();
}
```

Model: a MessageThread, with a OneToMany to Message. Message has a ManyToOne to user, named author. Message has a set of Likes (OneToMany), each Like has a liker (OneToMany) which is a User.

This gets a full thread, with all messages and likes on those messages - except when someone blocked another user - then the messages and likes get filtered.

If you remove one of the filterMany-statements and change it to fetchAll(), it works fine. In the current form, it does this NullPointerException. Cause appears to be that bindParams is null. Also SQL is null, and subQuery.getGeneratedSql() also returns null.

```java
Exception in thread "main" java.lang.NullPointerException
	at com.avaje.ebeaninternal.server.expression.InQueryExpression.queryPlanHash(InQueryExpression.java:59)
	at com.avaje.ebeaninternal.server.expression.DefaultExpressionList.queryPlanHash(DefaultExpressionList.java:391)
	at com.avaje.ebeaninternal.server.querydefn.OrmQueryProperties.queryPlanHash(OrmQueryProperties.java:500)
	at com.avaje.ebeaninternal.server.querydefn.OrmQueryDetail.queryPlanHash(OrmQueryDetail.java:75)
	at com.avaje.ebeaninternal.server.querydefn.OrmQueryPlanKey.<init>(OrmQueryPlanKey.java:83)
	at com.avaje.ebeaninternal.server.querydefn.DefaultOrmQuery.createQueryPlanKey(DefaultOrmQuery.java:899)
	at com.avaje.ebeaninternal.server.querydefn.DefaultOrmQuery.prepare(DefaultOrmQuery.java:914)
	at com.avaje.ebeaninternal.server.core.OrmQueryRequest.prepareQuery(OrmQueryRequest.java:157)
	at com.avaje.ebeaninternal.server.core.DefaultServer.createQueryRequest(DefaultServer.java:1070)
	at com.avaje.ebeaninternal.server.core.DefaultServer.createQueryRequest(DefaultServer.java:1051)
	at com.avaje.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1399)
	at com.avaje.ebeaninternal.server.core.DefaultBeanLoader.executeQuery(DefaultBeanLoader.java:231)
	at com.avaje.ebeaninternal.server.core.DefaultBeanLoader.loadMany(DefaultBeanLoader.java:88)
	at com.avaje.ebeaninternal.server.core.DefaultServer.loadMany(DefaultServer.java:482)
	at com.avaje.ebeaninternal.server.loadcontext.DLoadManyContext.loadSecondaryQuery(DLoadManyContext.java:98)
	at com.avaje.ebeaninternal.server.loadcontext.DLoadContext.executeSecondaryQueries(DLoadContext.java:190)
	at com.avaje.ebeaninternal.server.core.OrmQueryRequest.executeSecondaryQueries(OrmQueryRequest.java:98)
	at com.avaje.ebeaninternal.server.query.CQueryEngine.findMany(CQueryEngine.java:319)
	at com.avaje.ebeaninternal.server.query.DefaultOrmQueryEngine.findMany(DefaultOrmQueryEngine.java:94)
	at com.avaje.ebeaninternal.server.core.OrmQueryRequest.findList(OrmQueryRequest.java:320)
	at com.avaje.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1411)
	at com.avaje.ebeaninternal.server.core.DefaultServer.findUnique(DefaultServer.java:1169)
	at com.avaje.ebeaninternal.server.querydefn.DefaultOrmQuery.findUnique(DefaultOrmQuery.java:1167)
	at org.avaje.ebean.typequery.TQRootBean.findUnique(TQRootBean.java:903)
```

**Comment by pieterbos** [comment]
By the way, the Ebean to hibernate 5 comparison showed that hibernate 5 still hasn't got proper fetching of collections. They do OK if you do eager loading, but all they can do dynamically is joins. This makes every serious application in hibernate a performance problem waiting to happen, and limit/offset fetching really annoying (requiring a query to fetch the IDs, then the entities, OR a full fetch in memory of the entire query result). And even then Hibernate is slow.

The two nested filtermany's on the other hand, have a very easy workaround :)

**Comment by rbygrave** [comment]
Interesting thanks.  Yes the "Load Context" is the unsung hero of ORM - its
nice that we have a good one in Ebean (and it's proving useful in the
ElasticSeach use case as well which will be rolling out shortly).

It did prompt me to have a look at some of the changes in Hibernate 5. I
knew they were headed down the enhancement path (away from dynamic proxies)
and it looks they are putting some things in place that could take them to
a happier place.

For this one yes, I'll get to it shortly :)

Cheers, Rob.

**Comment by rbygrave** [comment]
What version are you using to reproduce this?

I was not able to reproduce yet but instead found: https://github.com/ebean-orm/avaje-ebeanorm/issues/601

**Comment by pieterbos** [comment]
I use version 7.1.1 - the very latest, as far as I know.

I just gave you access to the private repository https://github.com/nedap/orm-evaluation. Not yet public, I want to make a generic ORM-independent part and an ORM-specific implementation before I release this.

The example_for_rob_bygrave branch has the bug, running on a h2 database. Make sure you're running JDK8, then do:

```sh
git clone git@github.com:nedap/orm-evaluation.git
cd orm-evaluation
git checkout example_for_rob_bygrave
cd ebean
./gradlew clean run
```

Configured to run with a h2 in memory database, so it should just work - and did on a coworkers' computer.

After a few seconds it should log the Exception. The code that causes it is in ./src/main/java/com/nedap/orm/ebean/models/finder/MessageThreadFinder.java, method byIdFullyFetched.

**Comment by pieterbos** [comment]
I think it might have something to do with the fact that i used two filtermany-calls, both with subqueries, and that the second subquery did not get compiled into a query (query plan?).

**Comment by rbygrave** [comment]
So this reproduces in 7.1.1 but not in the 7.2.1-SNAPSHOT and the fix for #601 has fixed this.

Note that BulkProcessor line 90 to 92 need to be removed (as the Ebean.currentTransaction() returns null - as it is ended at line 87 or 85 via end() or commit().) ... so we need to remove lines 90 to 92.

**Comment by rbygrave** [comment]
Checking logs, an example of summary logging for a MessageThread and the 2 subsequent filterMany queries is:

FindMany type[MessageThread] origin[B9Ml6Z.CxEVUw.CH_eY8] exeMicros[255] rows[1] name[] predicates[t0.id = ? ] bind[5]

FindMany mode[+query] type[Message] origin[B9Ml6Z.CxEVUw.CH_eY8] load[path:messages size:1] exeMicros[2416] rows[107] name[] predicates[(t0.thread_id) in (?)  and  (t0.author_id) not in (select ...

FindMany mode[+query] type[Like] origin[B9Ml6Z.CxEVUw.CH_eY8] load[path:messages.likes size:10] exeMicros[4892] rows[231] name[] predicates[(t0.liked_id) in (?,?,?,?,?,?,?,?,?,?)  and  (t0.liker_id) not in (select ...

... so yes that is all looking good.

**Comment by rbygrave** [comment]
Closing as fixed in 7.2.1 (which will be released shortly - a day or 2).

**Comment by pieterbos** [comment]
Thanks a lot! :)
